File: Java/dependency-injection/changing-behavior-through-composition/src/main/java/com/github/grzesiek_galezowski/ChangingBehaviorThroughComposition/Alarms/ParallelCompoundAlarm.java
package com.github.grzesiek_galezowski.ChangingBehaviorThroughComposition.Alarms;

import com.github.grzesiek_galezowski.ChangingBehaviorThroughComposition.Alarm;

import java.util.Arrays;

public class ParallelCompoundAlarm implements Alarm {
  private final Alarm[] _alarms;

  public ParallelCompoundAlarm(Alarm... alarms) {
    _alarms = alarms;
  }

  public void trigger() {
    Arrays.stream(_alarms).parallel().forEach(alarm -> alarm.trigger());
  }

  public void disable() {
    Arrays.stream(_alarms).parallel().forEach(alarm -> alarm.trigger());
  }
}

File: Java/dependency-injection/changing-behavior-through-composition/src/main/java/com/github/grzesiek_galezowski/ChangingBehaviorThroughComposition/Alarms/ParallelCompoundAlarm.cs
using System.Threading.Tasks;

namespace ChangingBehaviorThroughComposition
{
  public class ParallelCompoundAlarm : Alarm
  {
    private readonly Alarm[] _alarms;

    public ParallelCompoundAlarm(params Alarm[] alarms)
    {
      _alarms = alarms;
    }

    public void Trigger()
    {
      Parallel.ForEach(_alarms, alarm => alarm.Trigger());
    }

    public void Disable()
    {
      Parallel.ForEach(_alarms, alarm => alarm.Trigger());
    }
  }
}